### PR TITLE
[CI][XPU] Add initial PR CI workflow for XPU device tests

### PR DESCRIPTION
**Pull request by pavansivaram**
## Motivation

Seed the XPU CI story for sglang-omni. Nothing today verifies per-PR that XPU-path changes still build and pass unit tests on Intel XPU hardware — regressions land unnoticed until a nightly or a manual repro. This PR wires the first per-PR XPU workflow with a single test (`tests/unit_test/xpu/test_device_layer.py`) so the scaffolding is exercised end-to-end before we grow the suite.

## What lands

- `.github/workflows/omni-xpu-ci.yaml` — new workflow.
- `docker/xpu.Dockerfile` — small delta so the image can be built in place on the runner.

## Workflow — key decisions

- **`dorny/paths-filter` job, not `on.pull_request.paths`.** If XPU CI ever becomes a required status, an `on.paths` filter would leave unrelated PRs pending forever; a filter job skips cleanly (skip = pass).
- **Cheap gate on GitHub-hosted, heavy build on self-hosted.** `run-ci` label + draft/mergeability checks run on `ubuntu-latest`; only then is the `intel-omni-bmg` runner claimed.
- **`preflight` pins `exact_sha`.** A force-push mid-run cannot split the gate and the test across two commits.
- **In-place `docker build`, no registry pull.** Every push exercises `docker/xpu.Dockerfile` itself; no companion publish workflow, no image/test drift.
- **`pytest` installed at test time, not baked into the image.** Test-time deps stay out of the image.
- **Cleanup SIGTERMs sglang before `docker rm -f`, then prunes builder/image cache.** Hard-killing a container with a live GPU context triggers a GT reset on Battlemage. Prunes stop cancelled builds from accumulating on the runner.

## Dockerfile delta — key decisions

- **Stdlib venv (`python3.12-venv` + `python3 -m venv /opt/venv`).** The base ships Python 3.12 without pip. An earlier iteration bootstrapped via `curl | sh` of the `uv` installer; a silent curl failure would ship a broken image with no `pip` on PATH, so this switched to the first-party stdlib venv.
- **Render group created up front (`RENDER_GID=110`).** The 2026.0.0 base dropped the `render` group; without it, the workflow's `docker run --group-add render` exits 125.

## Runtime prerequisites (fleet-side, not enforced by the workflow)

- self-hosted runner labeled `[self-hosted, intel-omni-bmg]` with `/dev/dri` exposed and the runner user in the `render` / `video` groups;
- docker daemon reachable to the runner user.

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests. (Wires the already-existing `tests/unit_test/xpu/test_device_layer.py` into CI.)
- [x] Update documentation / docstrings / example tutorials as needed. (Workflow header + Dockerfile comments carry the rationale.)

### Followups tracked separately

- Broaden the XPU suite beyond `test_device_layer.py`.
- Nightly counterpart (cron, wider matrix) — sglang's `nightly-test-intel.yml` is the template.
- `xpu-ci-job-monitor` analogue for analytics.

### CODE REVIEW

**Comment by jianan-gu** [comment]
@pavansivaram pls help resolve conflicts., as the https://github.com/sgl-project/sglang-omni/pull/1767 fix is merged, we shall cover it in CI. Thanks.

File: .github/workflows/omni-xpu-ci.yaml [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            docker exec "${CONTAINER_NAME}" bash -c '
+              pkill -TERM -f "sglang|run_suite|python3.*test_" 2>/dev/null || true
+              for _ in $(seq 1 30); do
+                pgrep -f "sglang::|sglang.launch_server" >/dev/null || break
+                sleep 1
+              done
+              pkill -KILL -f "sglang|run_suite" 2>/dev/null || true
+            ' || true
+          fi
+          docker rm -f "${CONTAINER_NAME}" || true
```
Comment (jianan-gu) [inline .github/workflows/omni-xpu-ci.yaml:239]: Do we also need to consider clean up things like build cache via:
```
docker builder prune -f 
docker image prune -f
```
As PR close/rebase triggers cancelling, but the cache is still there, which not clean for maintaining. Besides, we may also refer to the current practice in: https://github.com/sgl-project/sglang-omni/blob/main/.github/scripts/cleanup_ci_pr_home.sh

File: .github/workflows/omni-xpu-ci.yaml [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+      - name: Build XPU docker image
+        # Heaviest step: torch+xpu wheels, sglang clone + build, sglang-omni
+        # editable install. Runs the Dockerfile end-to-end so this workflow
+        # doubles as Dockerfile CI.
+        timeout-minutes: 90
+        shell: bash
+        run: |
+          set -euxo pipefail
+          docker build -f docker/xpu.Dockerfile -t "${IMAGE_TAG}" .
```
Comment (jianan-gu) [inline .github/workflows/omni-xpu-ci.yaml:175]: correct PR description as we are doing inplace docker building instead of from pre-built hub.

File: .github/workflows/omni-xpu-ci.yaml [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            "${IMAGE_TAG}" \
+            sleep infinity
+
+      - name: Show XPU state
+        shell: bash
+        run: |
+          docker exec "${CONTAINER_NAME}" python -c "import torch; print('torch', torch.__version__, 'xpu avail:', hasattr(torch, 'xpu') and torch.xpu.is_available())"
+          docker exec "${CONTAINER_NAME}" bash -c "command -v sycl-ls >/dev/null && sycl-ls || true"
+
+      - name: Run test_device_layer.py
```
Comment (arathi-hlab) [inline .github/workflows/omni-xpu-ci.yaml]: stage name should be something like Run Unit Test instead of  file name (.py)in stage

File: .github/workflows/omni-xpu-ci.yaml [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+# (torch+xpu wheels, sglang clone + install, sglang-omni install). In
+# exchange there is no dependency on a companion release workflow, and
+# every test run exercises the Dockerfile itself.
+#
+# Runtime prerequisites (fleet-side, not enforced by this workflow):
+#   * self-hosted runner labeled [self-hosted, intel-omni-bmg] with /dev/dri
+#     exposed, docker daemon reachable to the runner user, and the runner
+#     user in the render/video groups.
+
+on:
```
Comment (jianan-gu) [inline .github/workflows/omni-xpu-ci.yaml:32]: One more thing to check:
With `on.pull_request.paths`, a PR with no changes on below paths never triggers this workflow. If "XPU CI" is later made a required status check, those PRs stay pending forever and can't merge.

Suggestion: instead of filtering at the workflow level, always trigger and gate the heavy job with a `check-changes` job (`dorny/paths-filter`) + `if: `on the job.
Similar in: https://github.com/sgl-project/sglang/blob/main/.github/workflows/pr-test-xpu.yml

Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml:32]: Done

File: .github/workflows/omni-xpu-ci.yaml (hunk 3 above)
Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml]: Done

File: .github/workflows/omni-xpu-ci.yaml (hunk 2 above)
Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml:175]: Done

File: .github/workflows/omni-xpu-ci.yaml (hunk 1 above)
Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml:239]: Done

**Comment by pavansivaram** [comment]
I have fixed the comments and cleaned up the PR. Please review and approve if okay

File: docker/xpu.Dockerfile [hunk 5]
```diff
@@ -28,8 +28,20 @@ ARG GMM_VERSION=22.10.0
...
+ARG RENDER_GID=110
+RUN getent group render >/dev/null \
+    || groupadd --system --gid ${RENDER_GID} render \
+    || groupadd --system render
+
+ENV VIRTUAL_ENV=/opt/venv
+RUN python3 -m venv ${VIRTUAL_ENV}
+ENV PATH="${VIRTUAL_ENV}/bin:${PATH}"
+
+# Minors differ on purpose: the XPU channel ships no torchaudio 2.12+xpu.
```
Comment (siju-samuel) [inline docker/xpu.Dockerfile]: remove this comment

**Review by siju-samuel — APPROVED** [review]

**Review by jianan-gu — APPROVED** [review]

**Comment by jianan-gu** [comment]
Hi, @Ratish1 , could you please help also take a look? Thanks!

File: .github/workflows/omni-xpu-ci.yaml [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        if: always()
+        shell: bash
+        run: |
+          set -eux
+          # SIGTERM sglang before docker rm -f to drain GPU context and
+          # avoid a GT reset on Battlemage (learned from upstream sglang's
+          # pr-test-xpu.yml cleanup step).
+          if docker ps -a --format '{{.Names}}' | grep -qx "${CONTAINER_NAME}"; then
+            docker exec "${CONTAINER_NAME}" bash -c '
+              pkill -TERM -f "sglang|run_suite|python3.*test_" 2>/dev/null || true
```
Comment (Ratish1) [inline .github/workflows/omni-xpu-ci.yaml]: this `pkill -f` matches the enclosing `bash -c` because its argv contains `sglang`, so the shell exits before the drain loop. use self-excluding patterns such as `[s]glang`, `[r]un_suite`, and `[t]est_` in both `pkill` and `pgrep`.

File: .github/workflows/omni-xpu-ci.yaml [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                sleep 1
+              done
+              pkill -KILL -f "sglang|run_suite" 2>/dev/null || true
+            ' || true
+          fi
+          docker rm -f "${CONTAINER_NAME}" || true
+          docker rmi -f "${IMAGE_TAG}" || true
+          # Reap build cache / dangling images left behind when a prior push
+          # to this PR was cancelled mid-build; without this they accumulate.
+          docker builder prune -f || true
```
Comment (Ratish1) [inline .github/workflows/omni-xpu-ci.yaml]: this deletes the shared BuildKit cache after every successful run; observed runs removed 25–64 GB and forced the following 21–35 minute rebuilds. keep bounded cache and prune only stale or over-budget entries; the run-specific image cleanup above can remain.

File: .github/workflows/omni-xpu-ci.yaml [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+  checks: read
+  pull-requests: read
+
+jobs:
+  check-changes:
+    name: xpu-ci-changes
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    outputs:
+      xpu: ${{ steps.filter.outputs.xpu }}
```
Comment (Ratish1) [inline .github/workflows/omni-xpu-ci.yaml]: `workflow_dispatch` on `main` can make this filter inspect only the latest commit, so a manual run may stop after `xpu-ci-changes` when that commit does not touch an XPU path. force this output to true for manual dispatch and keep path filtering for pull requests.

File: .github/workflows/omni-xpu-ci.yaml (hunk 6 above)
Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml]: Modified the logic as suggested

File: .github/workflows/omni-xpu-ci.yaml (hunk 7 above)
Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml]: Removed `docker builder prune -f || true`

File: .github/workflows/omni-xpu-ci.yaml (hunk 8 above)
Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml]: Modified as below
`xpu: ${{ github.event_name == 'workflow_dispatch' && 'true' || steps.filter.outputs.xpu }}`

File: .github/workflows/omni-xpu-ci.yaml [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+            --name "${CONTAINER_NAME}" \
+            --device /dev/dri \
+            --group-add video \
+            --group-add render \
+            --shm-size 32g \
+            -v /dev/shm:/dev/shm \
+            "${IMAGE_TAG}" \
+            sleep infinity
+
+      - name: Show XPU state
```
Comment (yxs) [inline .github/workflows/omni-xpu-ci.yaml]: This step better be assert, not print. With a broken XPU stack the platform silently falls back to CPU and the device-agnostic tests still pass, keeping CI green.

**Review by Ratish1 — APPROVED** [review]
LGTM after @yxs  review comment is addressed

File: .github/workflows/omni-xpu-ci.yaml [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+              pkill -KILL -f "[s]glang|[r]un_suite" 2>/dev/null || true
+            ' || true
+          fi
+          docker rm -f "${CONTAINER_NAME}" || true
+          docker rmi -f "${IMAGE_TAG}" || true
+          # Reap dangling images left behind when a prior push was cancelled
+          # mid-build. NOT builder prune -f -- that evicts the shared
+          # BuildKit cache (25-64 GB observed) and forces a 20+ min rebuild
+          # on the next run.
+          docker image prune -f || true
```
Comment (yxs) [inline .github/workflows/omni-xpu-ci.yaml:245]: Just a reminder, with builder prune gone, nothing in this workflow bounds BuildKit cache growth, removing the image doesn't free build cache.
Fine if the runner's daemon-side BuildKit GC is enabled.

Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml:245]: We are planning to move docker build to separate job altogether and this workflow will be modified to use the latest available docker `docker pull`. So this wouldn't be an issue for now.

File: .github/workflows/omni-xpu-ci.yaml (hunk 9 above)
Comment (pavansivaram) [reply .github/workflows/omni-xpu-ci.yaml]: Done — turned the step into an assert. torch.xpu availability now fails the job with `sys.exit(1)` if false; renamed the step to `Verify XPU state` to reflect the new semantics. sycl-ls stays best-effort (diagnostic only).

**Comment by pavansivaram** [comment]
Hi @yxs 
I have made changes for the print to assertion in checking for XPU availability and noted the builder GC part.
Can you please review once and approve if all okay?

**Review by yxs — APPROVED** [review]
